cheatsheets/gleam-for-typescript-users.md

```diff
@@ -0,0 +1,827 @@
+between floats and integers in various ways including `rounding`, `floor`,
+`ceiling` and many more.
+
+Floats in gleam use float-specific operators like `+.` and `/.` instead of sharing `+` and `/`.
+
+```gleam
+1 / 2 // 0
+1.0 /. 2.0 // 0.5
+```
+
+### Arrays and Lists
+
+Arrays in TypeScript are allowed to have values of mixed types, but lists in Gleam can only have a single type.
+
+#### TypeScript
+
+TypeScript has arrays that support mixed types.
+
+```ts
+const arrayOfNumbers = [2, 3, 4];
+const arr2 = [1, "wow", true];
+const arr3: number[] = [1, 2, 3];
+const arr4: Array<number> = [1, 2, 3]; // eqvuialent
+
+const arr5: Array<string|number> = [1, "2", 3]; // Valid. Array of numbers and/or strings
+const arr6: number[] | string[] = [1, 2, 3]; // Valid. Array of numbers OR array of strings
+const arr7: number[] | string[] = ["one", "two", "three"]; // Valid
+const arr6: number[] | string[] = ["one", 2, "three"]; // Invalid
+```
+
+#### Gleam
+
+Gleam has a `cons` operator that works for lists destructuring and
+pattern matching. In Gleam lists are immutable so adding and removing elements
+from the start of a list is highly efficient.
+
+```gleam
+let list = [2, 3, 4]
+let list = [1, ..list]
+let [1, second_element, ..] = list
+[1.0, ..list] // compile error, type mismatch
+```
+
+### Tuples
+
+#### TypeScript
+
+TypeScript supports tuple types for arrays. They must be explicitly declared.
+
+```ts
+const a: [number, number] = [1, 2]; // valid
+const b: [number, number] = [1, 2, 3]; // not valid
+const c = [1, 2]; // inferred type is number[], NOT a tuple
+const d: [string, number] = ["wow", 42];
+```
+
+#### Gleam
+
+Tuples are very useful in Gleam as they're the only collection data type that
+allows mixed types in the collection.
+
+```gleam
+let my_tuple = #("username", "password", 10)
+let #(_, pwd, _) = my_tuple
+io.print(pwd) // "password"
+// Direct index access
+io.print(my_tuple.0) // "username"
+```
+
+
+### Destructuring Arrays and Tuples
+
+#### TypeScript
+
+TypeScript supports destructuring arrays with `const` or `let` statements.
+
+```ts 
+type NumberArray = number[];
+type NumberPair = [number, number]; // tuple
+
+const [a, b]: NumberArray = [1, 2]; // the type annotation here is not required
+console.log(a); // 1
+console.log(b); // 2
+
+const [c]: NumberPair = [3, 4]; // the type annotation here is not required
+console.log(c); // 3
+
+const [d, ...e] = [5, 6, 7, 8];
+console.log(d); // 5
+console.log(e); // [6, 7, 8]
+```
+
+#### Gleam
+
+Similar syntax is supported in Gleam for tuples. However, variadic destructuring is not.
+
+```gleam
+let #(a, b) = #(1, 2)
+
+io.debug(a) // 1
+io.debug(b) // 2
+
+let #(c, _) = #(3, 4)
+io.debug(c) // 3
+// _ ignores the 4
+```
+
+For lists, the `list.split` method may be used.
+
+```gleam
+import gleam/io
+import gleam/list
+
+pub fn main() {
+  let ints = [11, 22, 33]
+
+  let #(a, rest) = list.split(ints, 1)
```

Comment:
This is invalid code and will be rejected by the compiler.

We do not recommend using list.split like this either. It's not a useful pattern in Gleam so best not to confuse them by showing it.